### PR TITLE
Refactor towners.cpp

### PR DESCRIPTION
**Pull request by AJenbo**
This splits up the largest function in the game (TalkToTowner), and avoids searching for NPCs when talking to them (passing targets NPC instead), plus a bunch of other improvements to the code.

### CODE REVIEW

**Review by glebm — APPROVED** [review]

File: Source/towners.cpp [hunk 1]
```diff
@@ -103,1187 +98,1024 @@ char AnimOrder[6][148] = {
...
-	for (i = 0; i < NUM_TOWNERS; i++) {
-		switch (towners[i]._ttype) {
-		case TOWN_SMITH:
-			TownBlackSmith();
-			break;
-		case TOWN_HEALER:
-			TownHealer();
-			break;
+	int i = 0;
+	for (auto townerInit : TownerInitList) {
```
Comment (glebm) [inline Source/towners.cpp:987]: Avoid unnecessary copy here:

```suggestion
	for (auto &townerInit : TownerInitList) {
```

File: Source/towners.cpp [hunk 2]
```diff
@@ -103,1187 +98,1024 @@ char AnimOrder[6][148] = {
...
+		celPath = "Towners\\Girl\\Girls1.CEL";
 	}
+	towner._tAnimWidth = 96;
+	towner._tAnimOrder = -1;
+	LoadTownerAnimations(towner, celPath, 20, initData.dir, 6);
+	towner._tName = "Celia";
 }
 
-void InitFarmer()
+TownerInit TownerInitList[] = {
```
Comment (glebm) [inline Source/towners.cpp:303]: Can this be `const`?

File: Source/towners.cpp (hunk 1 above)
Comment (AJenbo) [reply Source/towners.cpp:987]: Oh didn't realize that, for some reason i though it just made it const...

File: Source/towners.cpp (hunk 2 above)
Comment (AJenbo) [reply Source/towners.cpp:303]: Defiantly, don't want to make the same mistake that the original does twice.

**Comment by AJenbo** [comment]
merged manually for no reason...
https://github.com/diasurgical/devilutionX/commit/f263e4e9d10502f7b71ee09fa70622dd902a98a1
https://github.com/diasurgical/devilutionX/commit/c98a1409f9fb0b8dc1a6cfedf8ff2136c9f10a61
https://github.com/diasurgical/devilutionX/commit/6321bf04d517cd62a6cd4e2273c8a38518847e2d
https://github.com/diasurgical/devilutionX/commit/e4e924eb72b27db9087c0c58ba3af6b157ae4f69
https://github.com/diasurgical/devilutionX/commit/4b577329f47b304c5c4f10713b2c13792880d315

**Pull request closed without merging**